### PR TITLE
Enhance REST API with Advanced Authentication and Endpoint Configuration

### PR DESCRIPTION
**Pull request by iklobato**
- Added `AbstractAuthentication` class for flexible authentication mechanisms.
- Implemented `DefaultJWTAuthentication` with dynamic SECRET_KEY generation.
- Updated `RestEndpoint` class to support custom authentication and configuration options.
- Added detailed docstrings and usage examples for `RestEndpoint`.
- Integrated advanced example of `RestEndpoint` configuration including custom authentication, HTTP method handling, and route exclusion.

These changes provide a more secure and flexible API framework, supporting advanced authentication and endpoint customization.

### CODE REVIEW

**Comment by iklobato** [comment]
Hi @MyGodItsFull0fStars this is one idea of the person usage
```
from authentication import DefaultJWTAuthentication

class ProductEndpoint(RestEndpoint):
    http_verbs = ['get', 'post', 'patch']  # Enable GET, POST, and PATCH methods
    http_exclude = ['put', 'delete']       # Exclude PUT and DELETE methods
    authentication = DefaultJWTAuthentication
```

The person can provide a single endpoint that depends or not of a model, and then, overwrite its methods to create the HTTP verbs from it. So on this example it will create GET, POST and PATCH automatically, and not create PUT and DELETE
Im also adding natively a class JWT authentication that handles automatically auhtentication if provided, with no additional libraries.

The person can also crate its own authentication method like
```
class MyCustomAuth(AbstractAuthentication):
    ...
```
and just need to provide generate_token and authenticate method, so its be called from the rest endpoint itself

**Comment by MyGodItsFull0fStars** [comment]
I skimmed the code changes and it looks very good so far! I will test it on my device tomorrow more thoroughly!

**Comment by iklobato** [comment]
That's awesome! My goal in the near future is to support startups in building their prototypes using this framework. The idea is to help them create the same solutions with fewer development hours. Most people don't want to spend too much time on simple tasks, so by keeping things straightforward, even managers will be able to create what they need on their own.



Thanks for the support!

File: lightapi/auth.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+        Returns:
+            Optional[Dict[str, str]]: A dictionary with user information if the token is valid,
+                                       otherwise None.
+        """
+        try:
+            payload = jwt.decode(token, self.SECRET_KEY, algorithms=["HS256"])
+            return payload
+        except jwt.ExpiredSignatureError:
+            # Handle expired token
```
Comment (MyGodItsFull0fStars) [inline lightapi/auth.py:195]: Yet missing, but I guess for now thats surely okay :)

File: lightapi/exceptions.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class MissingHandlerImplementationError(Exception):
+    """
+    Raised when a subclass does not implement the required handler for a specified HTTP verb.
+    """
+
+    def __init__(self, handler_name: str, verb: str) -> None:
+        super().__init__(
+            f"Missing implementation for {handler_name} required for HTTP verb: {verb}. "
```
Comment (MyGodItsFull0fStars) [inline lightapi/exceptions.py:8]: Is the trailing `blank` character intentional? If not please remove.

File: lightapi/rest.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+            Callable: The wrapped handler with optional authentication.
+        """
+        async def wrapped_handler(request: web.Request):
+            if self.authentication_class:
+                await self.authentication_class.authenticate_request(request)
+            return await handler(request)
+
+        return wrapped_handler
+
+    def configure(self) -> None:
```
Comment (MyGodItsFull0fStars) [inline lightapi/rest.py:183]: As it is an abstract class, the `@abstractmethod` decorator makes sense to be more declarative.

File: lightapi/lightapi.py [hunk 4]
```diff
@@ -1,3 +1,5 @@
+from typing import Dict, Type, Callable
```
Comment (MyGodItsFull0fStars) [inline lightapi/lightapi.py:1]: The imports should be grouped and sorted similar to:

```python
import logging
from typing import Callable, Dict, Type

from aiohttp import web
from sqlalchemy.exc import SQLAlchemyError

from lightapi.database import Base, engine
from lightapi.handlers import create_handler
from lightapi.rest import RestEndpoint
```

File: lightapi/auth.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+    Attributes:
+        SECRET_KEY (str): The secret key used for token encoding/decoding. This should be overridden
+                          by subclasses to set a specific secret key for the authentication mechanism.
+    """
+
+    SECRET_KEY: str
+
+    @abstractmethod
+    def authenticate(self, token: str) -> Type[NotImplementedError, Optional[Dict[str, str]]]:
```
Comment (MyGodItsFull0fStars) [inline lightapi/auth.py:30]: As `Type[...]` expects only one parameter, you need to wrap the inner part of the brackets with `Union` like `Type[Union[NotImplementedError, Optional[Dict[str, str]]]]`.

Maybe a type alias could make the code more readable as there is a lot happening at once.

File: lightapi/rest.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    - `POST /users/` for creating a new user.
+    - `GET /users/` for retrieving all users.
+    - `GET /users/{id}` for retrieving a user by ID.
+
+    If `authentication` is enabled, JWT validation will occur before handling each request.
+    """
+
+    tablename: str
+    http_verbs: List[str] = [
+        'post', 'get', 'put', 'delete', 'patch', 'options', 'head'
```
Comment (MyGodItsFull0fStars) [inline lightapi/rest.py:70]: IMO we should introduce a HTTP method constant class consisting of `StrEnum` or constant strings.
But that's my preference, as I am a sucker for enums and not reusing strings.

File: lightapi/auth.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+            str: The JWT token encoded with the user's ID.
+
+        Example:
+            jwt_auth = DefaultJWT()
+            token = jwt_auth.encode_token('user123')
+            print(f"Generated Token: {token}")
+        """
+        payload = {
+            'user_id': user_id,
+            'exp': datetime.datetime.utcnow() + self.EXPIRATION_DELTA
```
Comment (MyGodItsFull0fStars) [inline lightapi/auth.py:144]: FYI the method `utcnow` is marked as deprecated in more recent Python versions.

File: lightapi/auth.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
+from abc import ABC, abstractmethod
```
Comment (MyGodItsFull0fStars) [inline lightapi/auth.py:1]: The order of imports is recommended as:

```python
import datetime
import os
from abc import ABC, abstractmethod
from typing import Dict, Optional, Type, Union

import jwt
from aiohttp import web

```

**Review by MyGodItsFull0fStars — CHANGES_REQUESTED** [review]
Looks great to me! There were a few minor issues I have found, plus some suggestions, ideas and personal preferences.

Overall I really love the structure, it's clear, readable and easy to comprehend.

As said in one of the comments, if you prefer a broader Python support < 3.9 then I will change the _type hints_ in my commits to address this. 

I will mark the review as `Request changes` only because the imports should be sorted and grouped properly and the `configure` method of the `RestEndpoint` class likely needs a `@abstractmethod` decorator.

Have a nice day!
-Chris

File: lightapi/auth.py [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+import jwt
+from typing import Dict, Optional, Union, Type
+from aiohttp import web
+import datetime
+import os
+
+
+class AbstractAuthentication(ABC):
```
[the remark this answers is not in the document]
Comment (iklobato-w) [reply lightapi/auth.py:9]: Totally!

File: lightapi/auth.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+            print(f"Invalid or expired token: {str(e)}")
+
+        # Authenticating an HTTP request (assuming aiohttp request)
+        async def some_protected_endpoint(request):
+            user_info = await jwt_auth.authenticate_request(request)
+            return web.json_response({"message": "Authenticated", "user": user_info})
+
+    """
+
+    ALGORITHM = "HS256"
```
[the remark this answers is not in the document]
Comment (iklobato-w) [reply lightapi/auth.py:92]: Agree!

File: lightapi/rest.py (hunk 6 above)
Comment (iklobato-w) [reply lightapi/rest.py:70]: I think Enums will fit better here indeed

**Review by MyGodItsFull0fStars — COMMENTED** [review]
Awesome! Many of the changes you made resulted in leaner code, which is always nice!

I will be on vacation from tomorrow on, so while you can still assign PRs to me, I will only be available from 23th of September again, just to let you know!

**Comment by iklobato-w** [comment]
Hah no problem! Have a good trip!

I dont know from where you are, but if you use WhatsApp you can add me there +55 11 9 8434 8555

See u!

On 13 Sep 2024, at 18:18, Christian Bauer ***@***.***> wrote:



@MyGodItsFull0fStars commented on this pull request.

Awesome! Many of the changes you made resulted in leaner code, which is always nice!

I will be on vacation from tomorrow on, so while you can still assign PRs to me, I will only be available from 23th of September again, just to let you know!

**Pull request merged**